tensorcircuit/applications/optimization.py

```diff
@@ -0,0 +1,348 @@
+
+    return backend.real(loss)
+
+
+def QAOA_loss(
+    nlayers: int,
+    pauli_terms: Tensor,
+    weights: List[float],
+    params: List[float],
+    full_coupling: bool = False,
+    mixer: str = "X",
+) -> Any:
+    """
+    computes the loss function for the Quantum Approximate Optimization Algorithm (QAOA) applied to the Ising model.
+
+    :param nlayers: The number of layers in the QAOA ansatz.
+    :param pauli_terms: A list of Pauli terms, where each term is represented as a list of 0/1 series.
+    :param weights: A list of weights corresponding to each Pauli term.
+    :param params: A list of parameter values used in the QAOA ansatz.
+    :param full_coupling (optional): A flag indicating whether to use all-to-all coupling in mixers. Default is False.
+    :paran mixer (optional): The mixer operator to use. Default is "X". The other options are "XY" and "ZZ".
+    :return: The computed loss value.
+    """
+    c = QAOA_ansatz_for_Ising(
+        params, nlayers, pauli_terms, weights, mixer=mixer, full_coupling=full_coupling
+    )
+    # Obtain the quantum circuit using QAOA_from_Ising function
+
+    return Ising_loss(c, pauli_terms, weights)
+    # Compute the Ising loss using Ising_loss function on the obtained circuit
+
+
+def QUBO_QAOA(
+    Q: Tensor,
+    nlayers: int,
+    iterations: int,
+    vvag: bool = False,
+    ncircuits: int = 10,
+    init_params: Optional[List[float]] = None,
+    mixer: str = "X",
+    learning_rate: float = 1e-2,
+    callback: Optional[Optional[Callable[[List[float], float], None]]] = None,
+    full_coupling: bool = False,
+) -> Array:
+    """
+    Performs the QAOA on a given QUBO problem.
+    Adam optimizer from TensorFlow is used.
+
+    :param Q: The n-by-n square and symmetric Q-matrix representing the QUBO problem.
+    :param nlayers: The number of layers (depth) in the QAOA ansatz.
+    :param iterations: The number of iterations to run the optimization.
+    :param vvag (optional): A flag indicating whether to use vectorized variational adjoint gradient. Default is False.
+    :param ncircuits (optional): The number of circuits when using vectorized variational adjoint gradient.
+        Default is 10.
+    :param init_params (optional): The initial parameters for the ansatz circuit.
+        Default is None, which initializes the parameters randomly.
+    :paran mixer (optional): The mixer operator to use. Default is "X". The other options are "XY" and "ZZ".
+    :param learning_rate (optional): The learning rate for the Adam optimizer. Default is 1e-2.
+    :param callback (optional): A callback function that is executed during each iteration. Default is None.
+    :param full_coupling (optional): A flag indicating whether to use all-to-all coupling in mixers. Default is False.
+    :return params: The optimized parameters for the ansatz circuit.
+    """
+    if backend != get_backend("tensorflow"):
+        raise ValueError("`QUBO_QAOA` is designed for tensorflow backend.")
```

Comment:
why jax backend doesn't work? though it is ok that code in applications is backend dependent

Reply:
Okay i will just remove the backend check